src/frontend/src/optimizer/plan_node/mod.rs

```diff
@@ -51,11 +51,95 @@ use self::stream::StreamPlanRef;
 use self::utils::Distill;
 use super::property::{Distribution, FunctionalDependencySet, Order};
 
+/// A marker trait for different conventions, used for enforcing type safety.
+///
+/// Implementors are [`Logical`], [`Batch`], and [`Stream`].
+pub trait ConventionMarker: 'static + Sized {
+    /// The extra fields in the [`PlanBase`] of this convention.
+    type Extra: 'static + Eq + Hash + Clone + Debug;
+
+    /// Get the [`Convention`] enum value.
+    fn value() -> Convention;
+}
+
+/// The marker for logical convention.
+pub struct Logical;
+impl ConventionMarker for Logical {
+    type Extra = plan_base::NoExtra;
+
+    fn value() -> Convention {
+        Convention::Logical
+    }
+}
+
+/// The marker for batch convention.
+pub struct Batch;
+impl ConventionMarker for Batch {
+    type Extra = plan_base::BatchExtra;
+
+    fn value() -> Convention {
+        Convention::Batch
+    }
+}
+
+/// The marker for stream convention.
+pub struct Stream;
+impl ConventionMarker for Stream {
+    type Extra = plan_base::StreamExtra;
+
+    fn value() -> Convention {
+        Convention::Stream
+    }
+}
+
+/// The trait for accessing the meta data and [`PlanBase`] for plan nodes.
 pub trait PlanNodeMeta {
-    fn node_type(&self) -> PlanNodeType;
-    fn plan_base(&self) -> &PlanBase;
-    fn convention(&self) -> Convention;
+    type Convention: ConventionMarker;
+
+    const NODE_TYPE: PlanNodeType;
+
+    /// Get the reference to the [`PlanBase`] with corresponding convention.
+    fn plan_base(&self) -> &PlanBase<Self::Convention>;
+
+    /// Get the reference to the [`PlanBase`] with erased convention.
+    ///
+    /// This is mainly used for implementing [`AnyPlanNodeMeta`]. Callers should prefer
+    /// [`PlanNodeMeta::plan_base`] instead as it is more type-safe.
+    fn plan_base_ref(&self) -> PlanBaseRef<'_>;
+}
+
+// Intentionally made private.
+mod plan_node_meta {
+    use super::*;
+
+    /// The object-safe version of [`PlanNodeMeta`], used as a super trait of [`PlanNode`].
+    ///
+    /// Check [`PlanNodeMeta`] for more details.
+    pub trait AnyPlanNodeMeta {
+        fn node_type(&self) -> PlanNodeType;
+        fn plan_base(&self) -> PlanBaseRef<'_>;
+        fn convention(&self) -> Convention;
+    }
+
+    /// Implement [`AnyPlanNodeMeta`] for all [`PlanNodeMeta`].
+    impl<P> AnyPlanNodeMeta for P
+    where
+        P: PlanNodeMeta,
+    {
+        fn node_type(&self) -> PlanNodeType {
+            P::NODE_TYPE
+        }
+
+        fn plan_base(&self) -> PlanBaseRef<'_> {
+            PlanNodeMeta::plan_base_ref(self)
+        }
+
+        fn convention(&self) -> Convention {
+            P::Convention::value()
+        }
+    }
 }
```

Comment:
I am wondering whether we can expose a different property fetching interface for `PlanRef` to let developers know which (logical, batch or stream) properties they want to access explicitly. For example, `plan_ref.batch_property().unwrap()` to access the batch property with an explicit unwrap.

Reply:
Because accessing a property in `PlanRef` could panic underlying potentially might require a very high code coverage rate to ensure them work well. Exposing an `Option` or `Result` can help developers realize what are they doing right now and the potential result. At least we can throw an Error instead of panicking the whole system.

Reply:
> let developers know which (logical, batch or stream) properties they want to access explicitly

Yeah, this is a good point. Several reasons for not doing this now:
- I want to make the changes minimal in this PR.
- We may first further explore if it's possible to constrain `PlanRef` to something like `PlanRef<Convention>`, to reduce the usages of accessing on the convention-erased plan.


> Because accessing a property in `PlanRef` could panic underlying potentially might require a very high code coverage rate to ensure them work well.

Agree.


> At least we can throw an Error instead of panicking the whole system.

That's not a big problem. We always catch the panic when handling a SQL statement, so panicking will not affect other active sessions.